testsuite/tasks.py

```diff
@@ -1,44 +1,53 @@
 # pylint: disable=missing-docstring
+''' Tasks module.
+'''
 import io
 
 from behave.configuration import Configuration
 from behave.formatter.base import StreamOpener
 from behave.runner import Runner
 from celery import Celery
-import flask_socketio
 
+from testsuite.application import create_app
+from testsuite.extensions import db, socketio
+from testsuite.models.testrun import TestRun
 
-app = Celery()
-app.config_from_object('testsuite.celeryconfig')
-socketio = flask_socketio.SocketIO(message_queue='redis://')
 
+# Create and configure celery task runner
+celery = Celery()
+celery.config_from_object('testsuite.celeryconfig')
 
-@app.task
-def run_tests(room, vendor, tags, override):
 
-    def on_snapshot(snapshot, plan):
-        event = {
-            'snapshot': snapshot,
-            'plan': plan,
-        }
-        socketio.emit('snapshot', event, room=room)
-
-    try:
-        output = io.StringIO()
-        output_stream = StreamOpener(stream=output)
-        config = Configuration(
-            outputs=[output_stream],
-            format=['json.chunked'],
-            on_snapshot=on_snapshot,
-            vendor=vendor,
-            override=override,
-            command_args=[],
-            tags=[','.join(tags)],
-        )
-        runner = Runner(config)
-
-        runner.run()
-    except Exception as err:  # pylint: disable=broad-except
-        socketio.emit('global_error', str(err), room=room)
-    finally:
-        socketio.emit('tests_complete', room=room)
+@celery.task
+def run_tests(room, vendor, tags, override):
+    app = create_app()
+    with app.app_context():
+        test_run = TestRun()
+        db.session.add(test_run)
+        db.session.commit()
+
+        def on_snapshot(snapshot, plan):
+            test_run.save_snapshot(snapshot, plan)
```

Comment:
It could fail if something in the JSON were unserialize-able. That can happen with `datetime` objects, but if we've gotten this far, everything should already be safe (and if it isn't, `socketio.emit` will fail).

I think, because of the non-critical nature of this app, I'd rather have everything fail, post an error message, and have the vendors complain to us, rather than making sure the tests run at any cost, and possibly silently losing "save report" functionality.